rgw: add new zone filter support for bucket notifications

9401adarsh opened the pull request:
this fix adds support for zone filter for bucket notifications. 

changes introduced: 
- add new struct `rgw_s3_zone_filter` for filtering based on zone, add relevant struct  methods and match function for zone filter
- add condition to match notification based on zone filter in `notification_match` method in `rgw_notify.cc`. 

Fixes: https://tracker.ceph.com/issues/68789
Signed-off-by: Adarsh Ashokan [dev.9401adarsh@gmail.com](mailto:dev.9401adarsh@gmail.com)


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

9401adarsh committed:
rgw: add negative and zone filter support for bucket notifications

Fixes:  https://tracker.ceph.com/issues/68788
        https://tracker.ceph.com/issues/68789
Signed-off-by: Adarsh Ashokan <dev.9401adarsh@gmail.com>

9401adarsh commented:
Hi @yuvalif, I think I have a way to make the zone filter not based on the negative filter PR. 

We can have the zone filter to have two lists -> an in list and an out list instead of a map.
I feel this makes it much more easier to understand, and removes the overhead of a map. 

Would you suggest making this change ?

yuvalif commented:
> Hi @yuvalif, I think I have a way to make the zone filter not based on the negative filter PR.
> 
> We can have the zone filter to have two lists -> an in list and an out list instead of a map. I feel this makes it much more easier to understand, and removes the overhead of a map.
> 
> Would you suggest making this change ?

i think it makes sense to use the negative filter in the API.
you can have the implementation use 2 lists (agree, it makes more sense)

9401adarsh commented:
> > Hi @yuvalif, I think I have a way to make the zone filter not based on the negative filter PR.
> > We can have the zone filter to have two lists -> an in list and an out list instead of a map. I feel this makes it much more easier to understand, and removes the overhead of a map.
> > Would you suggest making this change ?
> 
> i think it makes sense to use the negative filter in the API. you can have the implementation use 2 lists (agree, it makes more sense)

I'll modify the uptake to do the following then: 

For the XML payload,  the zone filter will be as follows - (an array of filter rules having Name and Type attributes, where Type is IN or OUT)

```
<Zones>
    <FilterRule>
        <Name></Name>
        <Type></Type>
    </FilterRule>
</Zones>

```

The struct to have two lists `in_list` and `out_list` which will be used to match zones accordingly.

yuvalif commented:
please update:  `examples/rgw/boto3/service-2.sdk-extras.json`

9401adarsh commented:
> > Hi @yuvalif, have added the change to the services.json file and rebased off the latest negative filter PR as discussed in the mail.
> 
> could you test manually according to: https://gist.github.com/yuvalif/89e15839950c485a01732285d5e43614 ?

Hi @yuvalif, I have added more comments here regarding the test case. Can you please have a look ?

yuvalif commented on src/test/rgw/bucket_notification/test_bn.py:
```diff
@@ -2724,6 +2723,96 @@ def key_negative_filter(endpoint_type, conn):
+                        'Events': ['s3:ObjectCreated:*'],
+                        'Filter': {
+                          'Zones': {
+                            'FilterRules': [{'Name': 'default', 'Type': 'OUT'}]
```
aren't these rules contradicting one another?

yuvalif replied:
ok. these are 2 separate notifications.

9401adarsh commented on src/test/rgw/bucket_notification/test_bn.py:
```diff
@@ -2724,6 +2723,96 @@ def key_negative_filter(endpoint_type, conn):
+                        'Events': ['s3:ObjectCreated:*'],
+                        'Filter': {
+                          'Zones': {
+                            'FilterRules': [{'Name': 'default', 'Type': 'IN'}]
```
get_config_zonegroup() gets the value from the config file. 

would you suggest adding a new entry in the config file to allow for modification based on setup ? 

I'll add a new param in the config file called zone_name - seeded value to be `default` - and use that to populate the notification in the test.

adamemerson commented:
jenkins test make check arm64

aza547 reviewed (commented):
I'm very late to the party here but had some thoughts on this after implementing something similar [myself](https://github.com/ceph/ceph/pull/64453/) and not realising this PR existed 🤦 you might not be interested in my comments given that but figured i would post them anyway on the off chance they are helpful.

As background it is my use case that currently doesn't work with bucket notifications v2 that drove the original request for this feature. However I think the original requirements were more complex than they needed to be, and for my use case I actually don't need the negative filtering at all.

The requirement for negative zone filtering was based on the assumption that the zone would be retrieved from the object (i.e if object is written to siteA, when it appears on siteB via multisite sync its "zone" is still "siteA"). 

`const std::string &zone_name = res.store->get_zone()->get_name();`

But actually what this PR does is filters on the local zone from the RadosStore - which is better. So simply filtering by Replication event and zone name fulfills my requirement.

That said, it's also nice to have negative filtering as a feature for other users I'm sure. Anyway as above, aware this is already approved so feel free to take my comments with a pinch of salt, and thanks for implementing 😃

ivancich commented:
There are some other failures as well. Would you mind evaluating them to see if they're related to this PR, @9401adarsh ?

FAIL: bucket_notification.test_bn.test_notification_push_kafka_multiple_brokers_override
FAIL: bucket_notification.test_bn.test_notification_push_kafka_multiple_brokers_append

Test run: https://qa-proxy.ceph.com/teuthology/anuchaithra-2025-07-17_05:48:28-rgw-wip-anrao1-testing-2025-07-16-1526-distro-default-smithi/8392689/teuthology.log

QA tracker: https://tracker.ceph.com/issues/72111

ivancich commented:
Here's another one that I wonder whether it's related. Would you check this one too, @9401adarsh and @yuvalif ?

It's a failure on **test_bucket_logging_key_filter_j** .

test run: https://qa-proxy.ceph.com/teuthology/anuchaithra-2025-07-17_05:48:28-rgw-wip-anrao1-testing-2025-07-16-1526-distro-default-smithi/8392679/teuthology.log

qa tracker: https://tracker.ceph.com/issues/72111

9401adarsh commented:
FAIL: bucket_notification.test_bn.test_notification_push_kafka_multiple_brokers_override
FAIL: bucket_notification.test_bn.test_notification_push_kafka_multiple_brokers_append


> It's a failure on **test_bucket_logging_key_filter_j** .

these three tests are unrelated to the PR. @yuvalif, can you please confirm from your end as well ?

yuvalif commented on doc/radosgw/s3/bucketops.rst:
```diff
@@ -661,20 +682,29 @@ Response is XML encoded in the body of the request, in the following format:
+                        <Type></Type>
                     </FilterRule>
                 </S3Tags>
+                <Zones>
```
S3 does not have the concept of "zone" - this is the reason for the different naming

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -19,18 +19,21 @@ void rgw_s3_key_filter::dump(Formatter *f) const {
     f->open_object_section("");
     ::encode_json("Name", "prefix", f);
     ::encode_json("Value", prefix_rule, f);
+    ::encode_json("Type", is_prefix_negative == true ? "OUT" : "IN", f);
```
@aza547 agree, your proposal is cleaner.

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -157,42 +255,65 @@ void rgw_s3_filter::dump_xml(Formatter *f) const {
+    // else continue
+    bool is_negative = filter.is_regex_negative == true; 
+    bool does_regex_match = std::regex_match(key, base_regex); 
+    if (!(is_negative ^ does_regex_match)) {
```
agree that `==` is more commonly used than `^` :-)
however, the `^` logic reflect the comment above more directly.
maybe be even more verbose:
```cpp
if ((is_negative && !does_regex_match) || (!is_negative && does_regex_match)) {
...
```

yuvalif commented:
@9401adarsh as we collect commenst on the negative filter in this PR, and also to avoid rebase work, maybe close the other PR, and keep this PR for both zone and negative filter?

yuvalif commented:
> FAIL: bucket_notification.test_bn.test_notification_push_kafka_multiple_brokers_override FAIL: bucket_notification.test_bn.test_notification_push_kafka_multiple_brokers_append
> 
> > It's a failure on **test_bucket_logging_key_filter_j** .
> 
> these three tests are unrelated to the PR. @yuvalif, can you please confirm from your end as well ?

the multiple broker kafka failures are known issue - unrelated to this PR
the bucket logging key filter failure is not seen elsewhere and probably *is* related to this PR

9401adarsh commented:
rebased off latest main.

yuvalif commented on src/test/rgw/bucket_notification/bntests.conf.SAMPLE:
```diff
@@ -2,6 +2,7 @@
 port = 8000
 host = localhost
 zonegroup = default
+zonename = default
```
for the test to pass in teuthology, the following code should be modified:
https://github.com/ceph/ceph/blob/main/qa/tasks/notification_tests.py#L294

to include the new "zonename" parameter

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -45,18 +54,28 @@ bool rgw_s3_key_filter::decode_xml(XMLObj* obj) {
   auto suffix_not_set = true;
   auto regex_not_set = true;
   std::string name;
+  std::string type; 
```
can you please rename this to "exclude"?
also, can this be boolean instead of string? do we have proper xml encode/decode options for booleans?

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -69,18 +88,27 @@ void rgw_s3_key_filter::dump_xml(Formatter *f) const {
     ::encode_xml("Name", "prefix", f);
     ::encode_xml("Value", prefix_rule, f);
+    if(is_prefix_negative) {
+      ::encode_xml("Exclude",  true, f); 
```
```suggestion
      ::encode_xml("Exclude",  is_prefix_negative, f); 
```

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -94,56 +122,134 @@ void rgw_s3_key_value_filter::dump(Formatter *f) const {
+    ::encode_json("Name", key_value_type.first, f);
+    ::encode_json("Value", key_value_type.second.first, f);
+    if(key_value_type.second.second) {
+      ::encode_json("Exclude", true, f);
```
```suggestion
      ::encode_json("Exclude", key_value_type.second.second, f);
```

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -69,18 +88,27 @@ void rgw_s3_key_filter::dump_xml(Formatter *f) const {
     ::encode_xml("Name", "suffix", f);
     ::encode_xml("Value", suffix_rule, f);
+    if(is_suffix_negative) {
+      ::encode_xml("Exclude",  true, f); 
```
same

yuvalif commented on src/rgw/rgw_s3_filter.cc:
```diff
@@ -94,56 +122,134 @@ void rgw_s3_key_value_filter::dump(Formatter *f) const {
 
   std::string key;
   std::string value;
+  std::string type;
```
what do we need "type" for?

9401adarsh commented:
Hi @yuvalif, have made the suggested changes and was able to verify the http tests on folio07.

9401adarsh commented:
jenkins test make check

9401adarsh commented:
jenkins test make check arm64

9401adarsh commented:
jenkins test api

9401adarsh commented:
jenkins test make check

yuvalif commented:
jenkins test make check

yuvalif commented:
unstale

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.

ceph-jenkins commented:
Thank you for your contribution.  Since you, the author, are not a member of the Ceph GitHub Org yet, our CI will not automatically run.  Any member of the Ceph Org may comment "ok  - to - test" (without the dashes) to allow the Jenkins jobs to run.
